src/support/modify.c

```diff
@@ -172,6 +178,172 @@ __modify_apply_one(WT_SESSION_IMPL *session, WT_CURSOR *cursor,
 	return (0);
 }
 
+/*
+ * __modify_apply_no_overlap --
+ *	Apply a single set of WT_MODIFY changes to a buffer, where the changes
+ * are in sorted order and none of the changes overlap.
+ */
+static int
+__modify_apply_no_overlap(WT_SESSION_IMPL *session,
+    WT_CURSOR *cursor, WT_MODIFY *entries, int nentries, bool *successp)
+{
+	WT_DECL_ITEM(tmp);
+	WT_DECL_RET;
+	WT_ITEM *value;
+	size_t n, vsize;
+	const uint8_t *from;
+	uint8_t *to;
+	int i;
+	bool sformat;
+
+	*successp = false;
+
+	value = &cursor->value;
+	sformat = cursor->value_format[0] == 'S';
+
+	/*
+	 * If the modifications are sorted and don't overlap, we need a second
+	 * buffer but we can do the modifications in a single pass.
+	 *
+	 * The requirement for ordering is unfortunate, but modifications are
+	 * performed "in order", and applications specify byte offsets based on
+	 * that. (In other words, byte offsets are cumulative, modifications
+	 * that shrink or grow the data affect subsequent modification's byte
+	 * offsets.) Byte offsets become hard if we re-order modifications.
+	 */
+	for (vsize = value->size, i = 0; i < nentries; ++i) {
+		/* Check if the entries are sorted and non-overlapping. */
+		if (i + 1 < nentries &&
+		    (entries[i + 1].offset <
+		    entries[i].offset + entries[i].size ||
+		    entries[i + 1].offset <
+		    entries[i].offset + entries[i].data.size))
+			return (0);
+
+		/*
+		 * If either the data size or replacement count extend past the
+		 * end of the current value, we have to deal with padding bytes.
+		 * Don't try to fast-path padding bytes; it's not common and
+		 * adds branches to the loop applying the changes.
+		 */
+		if (vsize < entries[i].offset + entries[i].size ||
+		    vsize < entries[i].offset + entries[i].data.size)
+			return (0);
+		vsize += (entries[i].data.size - entries[i].size);
+	}
+
+	/* Get a scratch buffer than can hold the result. */
+	WT_RET(__wt_scr_alloc(session, vsize, &tmp));
+
+	/*
+	 * Walk the list of modifications, taking from the original buffer
+	 * anywhere there are gaps. Don't bother checking for copies of 0
+	 * bytes, assume it's unlikely in order to get rid of the branches.
+	 */
+	from = value->data;
+	to = tmp->mem;
+	for (i = 0; i < nentries; ++i) {
+		WT_ASSERT(session,
+		    entries[i].offset >= WT_PTRDIFF(to, tmp->mem));
+		n = entries[i].offset - WT_PTRDIFF(to, tmp->mem);
+		WT_ASSERT(session,
+		    from + n <= (uint8_t *)value->data + value->size);
+
+		memmove(to, from, n);
+		to += n;
+		memmove(to, entries[i].data.data, entries[i].data.size);
+		to += entries[i].data.size;
+
+		from += n + entries[i].size;
+	}
+
+	/* Copy any data remaining in the original buffer. */
+	n = value->size - WT_PTRDIFF(from, value->data);
```

Comment:
When we copy the remaining data, we must be copying the trailing '\0' too? Unless that got replaced by a modify.
So if we restore the trailing null, there might be cases we are adding an extra '\0'?

Reply:
Nice catch, thanks! Fixed in a0336bc.